slam-libraries/viam-orb-slam3/orbslam_server_v1.cc

```diff
@@ -110,6 +110,84 @@ ::grpc::Status SLAMServiceImpl::GetPositionNew(
     return grpc::Status::OK;
 }
 
+::grpc::Status SLAMServiceImpl::GetPointCloudMap(
+    ServerContext *context, const GetPointCloudMapRequest *request,
+    GetPointCloudMapResponse *response) {
+    float max = 0;
+    float min = 10000;
+    std::vector<ORB_SLAM3::MapPoint *> actualMap;
+    {
+        std::lock_guard<std::mutex> lk(slam_mutex);
+        actualMap = currMapPoints;
+    }
+
+    if (actualMap.size() == 0) {
+        return grpc::Status(grpc::StatusCode::UNAVAILABLE,
+                            "currently no map points exist");
+    }
+
+    // take sparse slam map and convert into a pcd. Orientation of PCD
+    // is wrt the camera (z is coming out of the lens) so may need to
+    // transform.
+
+    std::stringbuf buffer;
```

Comment:
@JohnN193 I copied the pattern from here https://github.com/viamrobotics/slam/pull/9/files#diff-b68d96da1bc970406f0f4402811d582812550e7c8450a636d73540f0363c73d8R136 in `GetMap` & have a few questions:


1. When the original GetMap code was introduced james provided some feedback that this could be refactored to use sprintf, would this now be a good time to do that refactor?
2. Why are we creating both a `std::stringbuf` , and then creating a `std::ostream` from it & then writing to both the ostream in some parts of the code ([here](https://github.com/viamrobotics/slam/pull/139/files#diff-b68d96da1bc970406f0f4402811d582812550e7c8450a636d73540f0363c73d8R137) & to the stringbuf [here](https://github.com/viamrobotics/slam/pull/139/files#diff-b68d96da1bc970406f0f4402811d582812550e7c8450a636d73540f0363c73d8R182) in others? Ideally I'd like to only have a single variable which is written to, but please educate me if we actually need to have 2 different objects with pointers to each other for some reason. 


cc @tessavitabile for C++ expertise

Reply:
I was looking into ifstream and simialr functions while doing GetInternalState [carto] and while it may not always be the case, ostream takes in a buffer object (either a file buffer  or string buffer). To my knowledge the ostream is handling the addition of bytes to the buffer object.

You might be able to get away with just a stringbuf but the ostream helps optimize performance.

https://stackoverflow.com/questions/48728330/whats-the-mechanism-of-iostreams-with-buffer

https://cplusplus.com/reference/ostream/ostream/ostream/

Reply:
yeah refactoring here would be fine, should help clean stuff up